app/main/views/buyers.py

```diff
@@ -533,6 +533,22 @@ def cancel_brief(framework_slug, lot_slug, brief_id):
     if brief["status"] != "closed":
         abort(404)
 
+    if award_flow:
+        label_text = "Why didn't you award a contract for {}?"
+        previous_page_url = url_for(
+            '.award_or_cancel_brief',
+            framework_slug=brief['frameworkSlug'],
+            lot_slug=brief['lotSlug'],
+            brief_id=brief['id']
+        )
+    else:
+        label_text = None
```

Comment:
I thought about this. I wanted the delete form to have a default option and if it has a default option then doing it here would be dupplicating.

I could make the delete form default something really generic like 'Do you want to delete?' but then there's no need for the briefs kwarg.
Unless I create a delete form base that is only used as a subclass for the 2 'personalised' delete brief forms that accept brief?

Reply:
Hmm, tricky. How about making the label_text an optional kwarg on the form? So we only overwrite the default if it's supplied? Something like

`label_text = kwargs.get('label_text', "Why do you need to cancel {}?")` ?

Reply:
So I've got something like that in the form:
`label_text = kwargs.pop('label_text') or 'Why do you need to cancel {}?'`
which I'm changing to the more pythonic:
`label_text = kwargs.pop('label_text', 'Why do you need to cancel {}?')`

But I have to define `label_text` here because it's used in 3 places in the view. I've put a comment now about the default text in the view.

What you think?

Reply:
`label_text = kwargs.pop('label_text', 'Why do you need to cancel {}?')` doesn't work. It'll return None if the key exists regardless of the value.
I expected 
```
{'foo': None}.pop('foo', 'bar') == 'bar'
```
I got
```
{'foo': None}.pop('foo', 'bar') == None
```

Reply:
Any reason why we can't do `.get` on the kwargs instead (I'd hope that would return `bar`)?

Reply:
Just figured out what you're getting at. I'll admit it did feel a bit weird.

We instinctively wanted to preserve the signature of the underlying base class by passing `form` as the first argument to our forms. 
The problem was that in doing so we couldn't use kwargs for _our_ arguments because they would be mistaken for positional kwargs in the **kwargs when passed to the base classes __init__ method, unless we popped them out of the kwargs dict.

```
# This is our base class. It takes a form_data kwarg
class Form(object):
    def __init__(self, form_data='something'):
        print form_data    
        

# First thought:
class OurForm(Form):
    def __init__(self, brief_responses, *args, **kwargs):
        super(OurForm, self).__init__(*args, **kwargs)
        print 'do something with our {}'.format(brief_responses)

OurForm('our_brief_responses', 'our_form_data')
# > our_form_data
# > do something with our our_brief_responses

# But that ^ doesn't let us pass form in first like the parent class, to remedy that let's try:

class OurForm(Form):
    def __init__(self, *args, **kwargs):
        kwargs.get('brief_responses')
        super(OurForm, self).__init__(*args, **kwargs)
        print 'do something with our {}'.format(brief_responses) 

# > OurForm('our_form_data', brief_responses='our_brief_responses')
# > TypeError: __init__() got an unexpected keyword argument 'brief_responses'


# The base class doesn't like that kwarg
# OK TypeError, we'll just pop our kwarg and not pass it on through
class OurForm(Form):
    def __init__(self, *args, **kwargs):
        brief_responses = kwargs.pop('brief_responses')
        super(OurForm, self).__init__(*args, **kwargs)
        print 'do something with our {}'.format(brief_responses) 

OurForm('our_form_data', brief_responses='our_brief_responses')
# > our_form_data
# > do something with our our_brief_responses


# OK that works (This is where the AwardForm was at) but...
# The plot thickens, default arguments
class OurForm(Form):
    def __init__(self, brief_responses='some default brief_responses', *args, **kwargs):
        super(OurForm, self).__init__(*args, **kwargs)
        print 'do something with our {}'.format(brief_responses) 

OurForm('our_form_data', brief_responses='our_actual_brief_resposnes')
# > TypeError: __init__() got multiple values for keyword argument 'brief_responses'

# Another type error? Wait, it thinks our_form_data is brief_responses too because of the positional args!


# We also tried `pop` and `get` without `or` for the default:
class OurForm(Form):
    def __init__(self, *args, **kwargs):
        brief_responses = kwargs.pop('brief_responses', 'some default brief_responses')
        super(OurForm, self).__init__(*args, **kwargs)
        print 'do something with our {}'.format(brief_responses)

OurForm('our_form_data', brief_responses='our_brief_responses')
> # do something with our our_brief_responses
> # our_form_data

OurForm('our_form_data', brief_responses=None)
> # do something with our None
> # our_form_data

# Pop didn't work ^ because it just looks for the key in the dict, and if its value is None then so be it


# Hence 'do something with our None'
# Let's try get:
class OurForm(Form):
    def __init__(self, *args, **kwargs):
        brief_responses = kwargs.get('brief_responses', 'some default brief_responses')
        super(OurForm, self).__init__(*args, **kwargs)
        print 'do something with our {}'.format(brief_responses)

OurForm('our_form_data', brief_responses='our_brief_responses')
# > TypeError: __init__() got an unexpected keyword argument 'brief_responses'
OurForm('our_form_data', brief_responses=None)
# > TypeError: __init__() got an unexpected keyword argument 'brief_responses'

# Get didn't work ^ because it doesn't remove the kwarg, and in any case it behaves the same as pop 
# just looks for the key in the dict, and if its value is None then so be it


# The solution I had in my PR was to use `or` to compare what we were getting from popping the kwarg:
class OurForm(Form):
    def __init__(self, *args, **kwargs):
        brief_responses = kwargs.pop('brief_responses') or 'some default brief_responses'
        super(OurForm, self).__init__(*args, **kwargs)
        print 'do something with our {}'.format(brief_responses)

OurForm('our_form_data', brief_responses='our_brief_responses')
# > our_form_data
# > do something with our our_brief_responses
OurForm('our_form_data', brief_responses=None)
# > our_form_data
# > do something with our some default brief_responses

# Which worked but felt a little kludgy, why do we need to pop at all..? The answer is we don't:


# As long as we stop trying to preserve the signature of the base class and pass our _required_ arguments first
# we can circle back round to our first thought. No TypeErrors, no duplicate kwargs. We never touch form_data
# The form doesn't need a default in a kwarg for brief_responses or label_text, it can legitimately require those
class OurForm(Form):
    def __init__(self, brief_responses, *args, **kwargs):
        super(OurForm, self).__init__(*args, **kwargs)
        print 'do something with our {}'.format(brief_responses)

OurForm('our_brief_responses', 'our_form_data')
# > our_form_data
# > do something with our our_brief_responses

# Works!
# There we have it,, that was the thought processes and attempts behind crating the AwardOrCancelForm, CancelForm and my refactor of the AwardForm
# Thanks for pushing me to the conclusion! Sorry for the essay! I wanted to write it down for myself too.

Reply:
Awesome - thanks for the explanation, all makes sense 👍